File: src/main/java/net/zargor/afterlife/permissionssystem/SessionManagement.java
package net.zargor.afterlife.permissionssystem;


import net.zargor.afterlife.WebServer;
import net.zargor.afterlife.objects.Group;
import net.zargor.afterlife.objects.Session;
import org.apache.commons.lang3.time.DateUtils;

import java.util.*;
import java.util.stream.*;

/**
 * Session management for the http server
 */
public class SessionManagement {

    private final List<Session> sessions = new ArrayList<>();


    public SessionManagement() {
        new Thread(() -> {
            while (true) {
                try {
                    synchronized (sessions) {
                        Date date = new Date();
                        sessions.removeAll(sessions.stream().filter(session -> session.getExpireDate() <= date.getTime()).collect(Collectors.toList()));
                    }
                    Thread.sleep(300000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }, "Sessionthread").start();
    }

    public synchronized String addSession(String name, Group group) {
        Optional<Session> sessionOptional = sessions.stream().filter(s -> s.getName().equals(name)).findFirst();
        if (sessionOptional.isPresent())
            return null;
        String sID = generateSessionID();
        sessions.add(new Session(sID, name, group, DateUtils.addDays(new Date(), 1).getTime()));
        return sID;
    }

    public synchronized void remSession(String id) {
        Optional<Session> sessionOptional = sessions.stream().filter(s -> s.getId().equals(id)).findFirst();
        sessionOptional.ifPresent(sessions::remove);
    }

    public synchronized Session getSessionByName(String name) {
        Optional<Session> sessionOptional = sessions.stream().filter(s -> s.getName().equals(name)).findFirst();
        return sessionOptional.orElse(null);
    }

    public synchronized Session getSessionByID(String id) {
        Optional<Session> sessionOptional = sessions.stream().filter(s -> s.getId().equals(id)).findFirst();
        return sessionOptional.orElse(null);

    }

    public String createCookieString(String name, Group group, boolean secure) {
        return String.format("z-sID=%s; Expires=%s; HttpOnly; %s", addSession(name, group), DateUtils.addDays(new Date(), 1).toGMTString(), secure ? "Secure;" : "");
    }

    public String createCookieString(String name, String groupName, boolean secure) {
        return String.format("z-sID=%s; Expires=%s; HttpOnly; %s", addSession(name, WebServer.getInstance().getGroupManagement().getGroup(groupName)), DateUtils.addDays(new Date(), 1).toGMTString(), secure ? "Secure;" : "");
    }

    private String generateSessionID() {
        String s = (UUID.randomUUID().toString() + UUID.randomUUID().toString()).replace("-", "");
        if (sessions.stream().anyMatch(session -> session.getId().equals(s)))
            return generateSessionID();
        return s;
    }
}

File: src/main/kotlin/net/zargor/afterlife/web/SessionManagement.kt
package net.zargor.afterlife.web

import net.zargor.afterlife.web.objects.Group
import net.zargor.afterlife.web.objects.Session
import org.apache.commons.lang3.time.DateUtils
import java.util.*

/**
 * Session management for the http server
 */
class SessionManagement(val main: WebServer) {
    private val sessions : MutableMap<String, Session> = mutableMapOf()

    @Synchronized
    fun addSession(id : String, group : Group) : String?{
        if (sessions.containsKey(id))
            return null
        val sID = generateSessionID()
        sessions.put(sID, Session(id, group))
        return sID
    }

    @Synchronized
    fun remSession(id : String) {
        if (sessions.containsKey(id)) {
            sessions.remove(id)
        }
    }

    @Synchronized
    fun getSessionsByName(name : String) : Session? {
        for((_,v) in sessions){
            if(v.name == name){
                return v
            }
        }
        return null
    }

    @Synchronized
    fun getSessionsByID(id : String) : Session? {
        if(sessions.containsKey(id)){
            return sessions[id]
        }
        return null
    }

    private fun generateSessionID() : String {
        val s = UUID.randomUUID().toString().replace("-","")
        if (sessions.containsKey(s)) {
            return generateSessionID()
        }
        return s
    }

    @Synchronized
    fun createCookieString(name : String, group: Group, secure : Boolean) : String =
            "z-sID=${addSession(name
            , group)}" +
            "; Expires=${DateUtils.addDays(Date(), 1).toGMTString()}" +
            "; HttpOnly" +
            "; ${if (secure) "Secure;" else ""}"

    @Synchronized
    fun createCookieString(name : String,groupName: String, secure : Boolean) : String {
        val group = main.handler.groupM.getGroup(groupName) ?: throw IllegalAccessException("Group not found!")
        return "z-sID=${addSession(name
                , group)}" +
                "; Expires=${DateUtils.addDays(Date(), 1).toGMTString()}" +
                "; HttpOnly" +
                "; ${if (secure) "Secure;" else ""}"
    }

}
